### PR TITLE
fix: unlock proregtx collateral on error

### PR DESCRIPTION
**Pull request by panleone**
## Issue being fixed
Unlock the (external) collateral of a proregtx if creation fails for any reason.

## What was done?
See commit description

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

**Review by UdjinM6 — DISMISSED** [review]
utACK

File: src/rpc/evo.cpp [hunk 1]
```diff
@@ -746,41 +741,59 @@ static UniValue protx_register_common_wrapper(const JSONRPCRequest& request,
...
-        if (!pkhash) {
-            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("collateral type not supported: %s", ptx.collateralOutpoint.ToStringShort()));
+        bool unlockOnError = false;
+        {
+            LOCK(wallet->cs_wallet);
+            if (!wallet->IsLockedCoin(ptx.collateralOutpoint.hash, ptx.collateralOutpoint.n)) {
+                wallet->LockCoin(ptx.collateralOutpoint);
+                unlockOnError = true;
+            }
         }
```
Comment (PastaPastaPasta) [inline src/rpc/evo.cpp:751]: How about this to make it const
```suggestion
        const bool unlockOnError = [&]() {
            if (LOCK(wallet->cs_wallet); !wallet->IsLockedCoin(ptx.collateralOutpoint.hash, ptx.collateralOutpoint.n)) {
                wallet->LockCoin(ptx.collateralOutpoint);
                return true;
            }
        }();
```

File: src/rpc/evo.cpp [hunk 2]
```diff
@@ -746,41 +741,59 @@ static UniValue protx_register_common_wrapper(const JSONRPCRequest& request,
...
+                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("collateral key not in wallet: %s", EncodeDestination(txDest)));
+                }
+                SignSpecialTxPayloadByString(tx, ptx, key);
+                SetTxPayload(tx, ptx);
+                return SignAndSendSpecialTx(request, chainman, tx, fSubmit);
+            }
+        } catch (...) {
+            if (unlockOnError) {
+                LOCK(wallet->cs_wallet);
+                wallet->UnlockCoin(ptx.collateralOutpoint);
```
Comment (PastaPastaPasta) [inline src/rpc/evo.cpp:794]: nit
```suggestion
                WITH_LOCK(wallet->cs_wallet, wallet->UnlockCoin(ptx.collateralOutpoint);
```

File: src/rpc/evo.cpp [hunk 3]
```diff
@@ -746,41 +741,59 @@ static UniValue protx_register_common_wrapper(const JSONRPCRequest& request,
...
             }
-            SignSpecialTxPayloadByString(tx, ptx, key);
-            SetTxPayload(tx, ptx);
-            return SignAndSendSpecialTx(request, chainman, tx, fSubmit);
+            throw;
```
Comment (PastaPastaPasta) [inline src/rpc/evo.cpp:796]: why is there a naked throw?

File: src/rpc/evo.cpp [hunk 4]
```diff
@@ -746,41 +741,59 @@ static UniValue protx_register_common_wrapper(const JSONRPCRequest& request,
...
+                LegacyScriptPubKeyMan* spk_man = wallet->GetLegacyScriptPubKeyMan();
+                if (!spk_man) {
+                    throw JSONRPCError(RPC_WALLET_ERROR, "This type of wallet does not support this command");
+                }
+
+                CKey key;
+                if (!spk_man->GetKey(ToKeyID(*pkhash), key)) {
+                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("collateral key not in wallet: %s", EncodeDestination(txDest)));
+                }
+                SignSpecialTxPayloadByString(tx, ptx, key);
```
Comment (PastaPastaPasta) [inline src/rpc/evo.cpp:787]: nit
```suggestion
                if (CKey key; !spk_man->GetKey(ToKeyID(*pkhash), key)) {
                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("collateral key not in wallet: %s", EncodeDestination(txDest)));
                } else {
                    SignSpecialTxPayloadByString(tx, ptx, key);
                }
```

**Review by PastaPastaPasta — COMMENTED** [review]
Mostly looks good; a few nits; and one concern around naked throw?

File: src/rpc/evo.cpp (hunk 3 above)
Comment (knst) [reply src/rpc/evo.cpp:796]: bcs need to re-throw the exception which has been caught by `except` without changing logic of code. If throw is naked - it throws the last exception again.

But I'd suggest to write here a wrapper around Lock/Unlock same as 'lock_guard + mutex works'.

File: src/rpc/evo.cpp (hunk 4 above)
Comment (UdjinM6) [reply src/rpc/evo.cpp:787]: I disagree

Comment (PastaPastaPasta) [reply src/rpc/evo.cpp:787]: good thing it was a nit :P

Comment (UdjinM6) [reply src/rpc/evo.cpp:787]: yeah, I'm not sure which version I like more actually 🙈

Comment (knst) [reply src/rpc/evo.cpp:787]: I'd prefer original one, bcs throw is throw, else looks ugly for throw

Comment (knst) [reply src/rpc/evo.cpp:787]: That's also an old code, there's just changed amount of whitespace but code is same byte-to-byte

Comment (UdjinM6) [reply src/rpc/evo.cpp:787]: we could maybe write it like that to make it less ugly :) while still keeping `key` inside of the `if` scope
```suggestion
                if (CKey key; spk_man->GetKey(ToKeyID(*pkhash), key)) {
                    SignSpecialTxPayloadByString(tx, ptx, key);
                } else {
                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("collateral key not in wallet: %s", EncodeDestination(txDest)));
                }
```

Comment (UdjinM6) [reply src/rpc/evo.cpp:787]: > That's also an old code, there's just changed amount of whitespace but code is same byte-to-byte

Good point, makes it easier to review with `?w=1` indeed

Comment (panleone) [reply src/rpc/evo.cpp:787]: for the moment I'm leaving it as it is since I did not modify that lines

**Review by knst — APPROVED** [review]
utACK

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for squash merge

**Pull request merged**